pkg/controller/clusterpool/collections.go

```diff
@@ -0,0 +1,435 @@
+}
+
+// ByName returns the named claim from the collection, or nil if no claim by that name exists.
+func (c *claimCollection) ByName(claimName string) *hivev1.ClusterClaim {
+	claim, _ := c.byClaimName[claimName]
+	return claim
+}
+
+// Unassigned returns a list of claims that are not assigned to clusters yet. The list is sorted by
+// age, oldest first.
+func (c *claimCollection) Unassigned() []*hivev1.ClusterClaim {
+	return c.unassigned
+}
+
+// Assign assigns the specified claim to the specified cluster, updating its spec and status on
+// the server. Errors updating the spec or status are bubbled up. Returns an error if the claim is
+// already assigned (to *any* CD). Does *not* validate that the CD isn't already assigned (to this
+// or another claim).
+func (claims *claimCollection) Assign(c client.Client, claim *hivev1.ClusterClaim, cd *hivev1.ClusterDeployment) error {
+	if claim.Spec.Namespace != "" {
+		return fmt.Errorf("Claim %s is already assigned to %s. This is a bug!", claim.Name, claim.Spec.Namespace)
+	}
+	for i, claimi := range claims.unassigned {
+		if claimi.Name == claim.Name {
+			// Update the spec
+			claimi.Spec.Namespace = cd.Namespace
+			if err := c.Update(context.Background(), claimi); err != nil {
+				return err
+			}
+			// Update the status
+			claimi.Status.Conditions = controllerutils.SetClusterClaimCondition(
+				claimi.Status.Conditions,
+				hivev1.ClusterClaimPendingCondition,
+				corev1.ConditionTrue,
+				"ClusterAssigned",
+				"Cluster assigned to ClusterClaim, awaiting claim",
+				controllerutils.UpdateConditionIfReasonOrMessageChange,
+			)
+			if err := c.Status().Update(context.Background(), claimi); err != nil {
+				return err
+			}
+			// "Move" the claim from the unassigned list to the assigned map.
+			// (unassigned remains sorted as this is a removal)
+			claims.byCDName[claim.Spec.Namespace] = claimi
+			copy(claims.unassigned[i:], claims.unassigned[i+1:])
+			claims.unassigned = claims.unassigned[:len(claims.unassigned)-1]
+			return nil
+		}
+	}
+	return fmt.Errorf("Claim %s is not assigned, but was not found in the unassigned list. This is a bug!", claim.Name)
+}
+
+// SyncClusterDeploymentAssignments makes sure each claim which purports to be assigned has the
+// correct CD assigned to it, updating the CD and/or claim on the server as necessary.
+func (claims *claimCollection) SyncClusterDeploymentAssignments(c client.Client, cds *cdCollection, logger log.FieldLogger) {
+	invalidCDs := []string{}
+	claimsToRemove := []string{}
+	for cdName, claim := range claims.byCDName {
+		cd := cds.ByName(cdName)
+		if cd == nil {
+			logger.Errorf("couldn't sync ClusterDeployment %s to the claim assigned to it (%s): ClusterDeployment not found", cdName, claim.Name)
+		} else if err := ensureClaimAssignment(c, claim, claims, cd, cds, logger); err != nil {
+			logger.WithError(err).Errorf("couldn't sync ClusterDeployment %s to the claim assigned to it (%s)", cdName, claim.Name)
```

Comment:
Suggest generally sticking with structured logging instead of formatting directives.